### PR TITLE
Refactoring/search select

### LINKED ISSUE #682: Create generic search-select component from lecturer selection component
- extract component from components/course/edit/LecturerSection.vue

### PR DESCRIPTION
**Pull request by tobias-wiese**
# Description
- Closes #682
 

## Reason for this PR
- We want a reusable search-select component for our application

## Changes in this PR
- created a generic search-select component for string arrays (handling of inputs in the parent component)
- replaced lecturer selection in course form by the new component + implemented new handling of selected string values
- adapted e2e tests

## Type of change (remove all that don't apply)
- Refactoring
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows 
- Browser: Firefox

- Frontend: (remove all that don't apply)
  - [x] Development build

- Backend: (remove all that don't apply)
  - [x] deployed development build

# Checklist: (remove all that don't apply)

- [x] I added corresponding E2E tests (especially for bugfixes)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new linting warnings or console warnings
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)

Screenshot:
grafik

### CODE REVIEW

**Review by palkra — APPROVED** [review]
Looks good!

File: src/components/course/edit/sections/LecturerSection.vue [hunk 1]
```diff
@@ -124,38 +95,27 @@
...
-                    lecturerFound.value = true;
-                    currentLecturer.value = lecturer;
-                } else {
-                    lecturerFound.value = false;
-                    currentLecturer.value = {} as Lecturer;
-                }
-            });
+            function createOptionString(l: Lecturer): string {
+                return l.firstName + " " + l.lastName + " (@" + l.username + ")";
+            }
```
Comment (theovier) [inline src/components/course/edit/sections/LecturerSection.vue:111]:
```suggestion
            function createOptionString(lecturer: Lecturer): string {
                return `$(lecturer.firstName) $(lecturer.lastName) (@$(lecturer.username))`;
            }
```

File: src/components/course/edit/sections/LecturerSection.vue [hunk 2]
```diff
@@ -79,43 +65,28 @@
...
-                optionsShown.value = true;
-            }
-
-            function keyMonitor(event: KeyboardEvent) {
-                if (event.key === "Enter" && filteredLecturers.value[0]) {
-                    selectOption(filteredLecturers.value[0]);
+            watch(selectedOption, () => {
+                let lecId = selectedOption.value.substring(selectedOption.value.lastIndexOf("@") + 1, selectedOption.value.length - 1);
+                let tmp = lecturers.value.find((l) => l.username == lecId) as Lecturer;
+                if (tmp != undefined) {
```
Comment (theovier) [inline src/components/course/edit/sections/LecturerSection.vue:81]:
```suggestion
                if (tmp !== undefined) {
```

File: src/components/course/edit/sections/LecturerSection.vue [hunk 3]
```diff
@@ -79,43 +65,28 @@
...
-            function showOptions() {
-                optionsShown.value = true;
-            }
-
-            function keyMonitor(event: KeyboardEvent) {
-                if (event.key === "Enter" && filteredLecturers.value[0]) {
-                    selectOption(filteredLecturers.value[0]);
+            watch(selectedOption, () => {
+                let lecId = selectedOption.value.substring(selectedOption.value.lastIndexOf("@") + 1, selectedOption.value.length - 1);
+                let tmp = lecturers.value.find((l) => l.username == lecId) as Lecturer;
```
Comment (theovier) [inline src/components/course/edit/sections/LecturerSection.vue:80]:
```suggestion
                let tmp = lecturers.value.find((l) => l.username === lecId) as Lecturer;
```

File: src/components/course/edit/sections/LecturerSection.vue [hunk 4]
```diff
@@ -79,43 +65,28 @@
         emits: ["update:lecturerId"],
         setup(props: any, { emit }: any) {
             const lecturers = ref([] as Lecturer[]);
-            const newLecturerId = ref(props.lecturerId);
             const currentLecturer = ref({} as Lecturer);
             const lecturerFound = ref(false);
-            const input = ref(props.lecturerId);
             const optionsShown = ref(false);
```
Comment (theovier) [inline src/components/course/edit/sections/LecturerSection.vue:70]:
```suggestion
            
```

**Review by theovier — COMMENTED** [review]
### Idea to improve the workflow with the component:

Instead of passing an array of options, pass an **array of objects**, like follows:

`[
    {
        value: Object,
        display: String
    }, ...
]`

When we v-model the `selectedOption`, we can then extract exactly the information we expect the item to have instead of using substrings (see lecturer section). We could make use of something like `selectedOption.value.id` or `selectedOption.value.name` to retrieve information of the selected option directly. 

What do you think about it @upb-uc4/vue ?

**Review by theovier — APPROVED** [review]

**Pull request merged**